### PR TITLE
Evol n-ary association

### PR DESCRIPTION
**Pull request by fbats**
Creation of n-ary association.

### CODE REVIEW

File: plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java [hunk 1]
```diff
@@ -81,6 +90,81 @@ public String computeAssociationEndLabel(Association association) {
...
 
 	/**
+	 * Compute label forn_ary association edge.
+	 *
+	 * @param association
+	 * @param view
+	 *            edge
+	 * @return label
+	 */
+	public String computeAssociationNAryBeginLabel(Association association, DDiagramElement view) {
```
Comment (mbats) [inline plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java]: This method should just call a LabelServices#computeNAryAssociationBeginLabel method

File: plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java [hunk 2]
```diff
@@ -119,29 +219,70 @@ public Type findTypeByName(EObject object, String typeName) {
...
-			a.getOwnedEnds().add(newOne);
-		} else {
-			/*
-			 * we already have both property ends, we just need to set the type
-			 */
-			if (target.getType() == null) {
-				target.setType(b);
-			} else if (source.getType() == null) {
-				source.setType(b);
+	// private void fixAssociation(Association a, Type b) {
```
Comment (mbats) [inline plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java]: Delete unused code

File: plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java [hunk 3]
```diff
@@ -325,6 +575,28 @@ public Type getSourceType(Association association) {
...
+	 * Get the type of the association source end.
+	 * @param association association
+	 * @param diagram diagram
+	 * @return type of the source
+	 *
+	 */
+	public Type getSourceType(Association association, DDiagram diagram) {
+		final EList<DDiagramElement> elements = diagram.getDiagramElements();
+		// List semantic elements visible in diagram
+		final List<EObject> list1 = new ArrayList<EObject>();
```
Comment (mbats) [inline plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java]: list1? search or a better name

File: plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java [hunk 4]
```diff
@@ -545,10 +932,12 @@ public boolean sourceIsComposite(Association association) {
...
 	 * @param association
 	 *            Association
+	 * @param element
+	 *            Edge element
 	 * @return True if source is navigable
 	 */
-	public boolean sourceIsNavigable(Association association) {
-		final Property source = AssociationServices.INSTANCE.getSource(association);
+	public boolean sourceIsNavigable(Association association, DDiagramElement element) {
+		final Property source = getSourceEndNaryAssociation(association, element);
```
Comment (mbats) [inline plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java]: if the association is not an N-ary one we still go through the sourceIsNavigable method so the name of the getSourceEndNAryAssociation is weird

File: plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/AbstractDiagramServices.java [hunk 5]
```diff
@@ -878,6 +878,7 @@ public Element reconnectEdge(Element context, DEdge edgeView, EdgeTarget sourceV
 		}
 		reconnectService.setOldPointedElement(source);
 		reconnectService.setNewPointedElement(target);
+		reconnectService.setEdge(edgeView);
```
Comment (mbats) [inline plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/AbstractDiagramServices.java]: I don't undertsand what is the purpose of this function. Review the name of the method ?

File: plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/internal/services/ReconnectSwitch.java [hunk 6]
```diff
@@ -485,6 +503,16 @@ private boolean isReconnectable(Element element) {
 	}
 
 	/**
+	 * Set edge.
```
Comment (mbats) [inline plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/internal/services/ReconnectSwitch.java]: Explain what is the purppose of this method.

File: plugins/org.obeonetwork.dsl.uml2.design/description/uml2.odesign [hunk 7]
```diff
@@ -1754,23 +1751,50 @@
...
+          <style lineStyle="dash_dot" targetArrow="NoDecoration">
```
Comment (mbats) [inline plugins/org.obeonetwork.dsl.uml2.design/description/uml2.odesign]: Should be solid

File: plugins/org.obeonetwork.dsl.uml2.design/images/association.svg [hunk 8]
```diff
@@ -0,0 +1,68 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (mbats) [inline plugins/org.obeonetwork.dsl.uml2.design/images/association.svg]: The file is cut in south direction, recontribute a new one

File: plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java [hunk 9]
```diff
@@ -275,6 +461,56 @@ public Element getBaseClass(EObject stereotypeApplication) {
...
+	 */
+	public List<Association> getNAryAssociation(Package container, DDiagram diagram) {
+		final List<Association> result = new ArrayList<Association>();
+		final Collection<EObject> associations = getAssociationInverseRefs(diagram);
+		for (final EObject object : associations) {
+			final Association association = (Association)object;
+			// Association should be visible in self container
+			// At least one of the ends is visible in diagram
+			final EList<DDiagramElement> elements = diagram.getDiagramElements();
+			// check if at least more than 2 ends are displayed in diagram
```
Comment (mbats) [inline plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java]: Review the name

File: plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java [hunk 10]
```diff
@@ -275,6 +461,56 @@ public Element getBaseClass(EObject stereotypeApplication) {
...
+			final List<EObject> list1 = new ArrayList<EObject>();
```
Comment (mbats) [inline plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java]: Review the list name

File: plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java [hunk 11]
```diff
@@ -313,6 +549,20 @@ public Element getBaseClass(EObject stereotypeApplication) {
 		return classes;
 	}
 
+	private Property getSourceEndNaryAssociation(Association association, DDiagramElement view){
+		final DEdge edge = (DEdge)view;
+		final String sourceObject = edge.getSourceNode().toString();
```
Comment (mbats) [inline plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java]: Why using toString ?

**Comment by mbats** [comment]
Can't update in case of an nary association in the member end table in property view the navigable and owned checkboxes.

File: plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/internal/services/ReconnectSwitch.java [hunk 12]
```diff
@@ -84,14 +85,31 @@
...
 	 * {@inheritDoc}
 	 */
 	@Override
 	public Element caseAssociation(Association association) {
 		if (isReconnectable(association)) {
 			if (RECONNECT_SOURCE == reconnectKind) {
-				final Property source = AssociationServices.INSTANCE.getSource(association);
-				source.setType((Type)newPointedElement);
+				if (association.getMemberEnds().size() > 2) {
+					final String sourceObject = edge.getSourceNode().toString();
```
Comment (mbats) [inline plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/internal/services/ReconnectSwitch.java]: Why using toString ?

**Comment by mbats** [comment]
Fix commit message

**Comment by mbats** [comment]
Contribute documentation

File: plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java (hunk 2 above)
Comment (fbats) [inline plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java]: Done

File: plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java (hunk 3 above)
Comment (fbats) [inline plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java]: done

File: plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java (hunk 4 above)
Comment (fbats) [inline plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java]: done

File: plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/AbstractDiagramServices.java (hunk 5 above)
Comment (fbats) [inline plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/AbstractDiagramServices.java]: done

File: plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/internal/services/ReconnectSwitch.java (hunk 6 above)
Comment (fbats) [inline plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/internal/services/ReconnectSwitch.java]: done

File: plugins/org.obeonetwork.dsl.uml2.design/description/uml2.odesign (hunk 7 above)
Comment (fbats) [inline plugins/org.obeonetwork.dsl.uml2.design/description/uml2.odesign]: done

File: plugins/org.obeonetwork.dsl.uml2.design/images/association.svg (hunk 8 above)
Comment (fbats) [inline plugins/org.obeonetwork.dsl.uml2.design/images/association.svg]: done

File: plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java (hunk 10 above)
Comment (fbats) [inline plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java]: done

File: plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java (hunk 1 above)
Comment (mbats) [inline plugins/org.obeonetwork.dsl.uml2.design/src/org/obeonetwork/dsl/uml2/design/api/services/ClassDiagramServices.java]: Done

**Comment by mbats** [comment]
Close this PR as a new one was pushed : #812

**Pull request closed without merging**